sql/src/main/java/io/crate/executor/transport/task/DropTableTask.java

```diff
@@ -151,6 +162,54 @@ public void onFailure(Exception e) {
         });
     }
 
+    private CompletableFuture<Void> dropPrivilegesAndIngestRulesForTable(TableIdent tableIdent) {
+        CompletableFuture<Long> dropPrivilegesFuture = dropUserPrivilegesFor(tableIdent);
+        CompletableFuture<Long> dropIngestRulesFuture = dropIngestRulesForTarget(tableIdent);
+        return CompletableFuture.allOf(dropPrivilegesFuture, dropIngestRulesFuture);
+    }
+
+    private CompletableFuture<Long> dropUserPrivilegesFor(TableIdent tableIdent) {
+        return userManager.dropTablePrivileges(tableIdent.fqn()).whenComplete((r, t) -> {
+            if (t != null) {
+                logger.warn(
+                    String.format(Locale.ENGLISH, "Unable to drop existing privileges for table %s.", tableIdent.fqn()),
+                    t);
+            }
+        });
+    }
+
+    private CompletableFuture<Long> dropIngestRulesForTarget(TableIdent tableIdent) {
+        return dropIngestRules(tableIdent).whenComplete((r, t) -> {
+            if (t != null) {
+                logger.warn(
+                    String.format(Locale.ENGLISH, "Unable to drop existing ingest rules for table %s.", tableIdent.fqn()),
+                    t);
+            }
+        });
+    }
+
+    private CompletableFuture<Long> dropIngestRules(TableIdent tableIdent) {
+        MetaData metaData = clusterService.state().metaData();
+        IngestRulesMetaData ingestRulesMetaData = (IngestRulesMetaData) metaData.getCustoms().get(IngestRulesMetaData.TYPE);
+        if (ingestRulesMetaData == null) {
+            return CompletableFuture.completedFuture(0L);
+        }
+
+        Set<IngestRule> rulesForTargetTable = ingestRulesMetaData.getAllRulesForTargetTable(tableIdent.fqn());
+        if (rulesForTargetTable.size() == 0) {
+            return CompletableFuture.completedFuture(0L);
+        }
+
+        FutureActionListener<IngestRuleResponse, Long> listener =
+            new FutureActionListener<>(r -> (long) rulesForTargetTable.size());
+        Set<String> ruleNamesToDrop = new HashSet<>(rulesForTargetTable.size());
+        for (IngestRule ingestRule : rulesForTargetTable) {
+            ruleNamesToDrop.add(ingestRule.getName());
+        }
```

Comment:
all the above checks and computations must be done at the transport action, otherwise we're running into race conditions.